### DISCUSSION TITLE
Raspberry Pi Pico Host bulk transfers: Is there a better way?

### DISCUSSION THREAD

**Discussion by rppicomidi** [hathach/tinyusb#1261] [Q&A] [answered]
# Summary
The way I implemented bulk endpoints for the RP2040 MIDI Host (pull request #1219) is a hack. I describe at length what I did and why because I am hoping someone will tell me the hack is not required because there is a better way. If you have a better way, or can tell me I am on the right track, I would appreciate it.
# The Problem
The RP2040 USB IP in host mode provides dedicated hardware for each interrupt endpoint. However, it offers just one endpoint hardware block to use for all control transfers, bulk transfers, and isochronous transfers.  The host controller software controls how to share this endpoint hardware among concurrent functions. For a MIDI device, that is typically one bulk IN endpoint and one bulk OUT endpoint.

This is my understanding of the RP2040 Host hardware operation: Every time the RP2040 USB host software needs to start a bulk IN or bulk OUT transfer, it sets up the shared endpoint control register and shared endpoint buffer register, and then writes to bits in the SIE Control register to start the transfer. The endpoint hardware retries this transfer forever until it completes or there is an error. NAK is not an error.

This is fine for control IN transfers because they are always setup command and response. You know that IN transfers will either STALL, timeout or complete. This is also fine for control OUT and bulk OUT transfers because the device can't NAK forever. It either has to accept the data sometime, or it has to STALL or timeout. 

Bulk IN transfers are the problem. The host can request data from a MIDI keyboard. That keyboard may not send anything until the player presses a key. Meanwhile, the host can't send data to the keyboard. Some MIDI devices will always send something (even a buffer full of 0) in response to an IN token, but most will NAK until something happens that requires the device to send something (e.g., the user presses a key or button, turns a knob, moves a slider, etc., or enough time has passed for a new active sensing message or MIDI clock event event). The automatic NAK retry behaves like a bug in this case, not a feature.
# My Ideal Solution
Given how the hardware works, my ideal algorithm for sharing the host endpoint between bulk IN and bulk OUT transfers is:

1. Do forever:
2. Poll the device (start an IN transfer)
3. If the transfer completes, get the data. If there is a NAK, go on
4. If there is data to send, start an OUT transfer. If not go back to the beginning
5. When the transfer completes, go back to the beginning.

What I wanted for bulk IN was a way to turn off automatic retry on NAK and instead get an interrupt on NAK. I could find no way to do that with the hardware registers. There is no documented way to stop the NAK auto-retry. `STALL_REC` is an interrupts source, but `NAK_REC` status is not, so there is no way I could find for the host controller software to get an IRQ when an endpoint responds NAK.

# The Solution Needs a Hack

1. Do forever:
2. Attempt to poll the IN endpoint. If no endpoint is using the single control/bulk/isochronous endpoint (EPx), start an IN transfer. If another endpoint is using EPx, go on. But if this endpoint is still using EPx, it might be waiting "forever" for data. Attempt to Abort the IN transfer. Move on if the abort succeeded. Go back to the beginning if the abort failed.
3. If there is data to send, attempt to start an OUT transfer. If this or another endpoint is using EPx, go on. Otherwise start the OUT transfer. When it completes, the OUT endpoint will no longer be using EPx.
4. Go back to the beginning

# Aborting the IN Transfer
This is the hack. In pull request #1219 it is coded in file `src/portable/raspberrypi/rp2040/hcd_rp2040.c` in the function `hcd_edpt_clear_in_on_nak()`. The source code follows in case reading my words just confuses you. The function is entered if there is an attempt to poll the IN endpoint and the IN endpoint is busy.

## Text Description

1. If a NAK was received (`SIE_STATUS.NAK_REC` is set) and there is currently no full buffer or a completed transaction as shown in the `INTS` register, then go on. Otherwise, the abort failed. Exit this routine. IN transfer endpoint structure will still show EPx busy.
2. Clear the `NAC_REC` status bit and wait for either a new NAK or a completed transaction. That takes several steps:
    1. Disable the USB interrupts in case a transaction completes while we are waiting for a NAK. That is, save the current value of the `INTE` register in `temp_inte` and then set `INTE` to 0.
    2. Clear the `NAK_REC` status bit
    3. Read the raw interrupt status from the `INTR` register and mask it with `temp_inte`.
    4. In a tight loop, wait for the NAK status bit to be set or for any of the masked `INTR` bits to be set.
    5. If the loop exits with NAK status clear or with masked `INTR` not 0, then restore the `INTE` register and return; a transfer completed or an error occurred, so it is not possible to stop the current IN transfer here. Otherwise, go on.
    6. If we got here, then the device NAK'd the IN transfer. Stop the IN transfer by setting the `STOP_TRANS` bit in the `SIE_CTRL` register. My observations show this shuts off the automatic IN transfer retry on NAK. This test and register write should probably be in a critical section because the transfer has to be stopped before the automatic retry kicks in. If another interrupt source triggers before the transfer is stopped, the automatic retry might start again. Right now, there is no critical section in the code.
    7. Clean up state as if the IN transfer completed:
        1. Clear the `NAK_REC` bit in `SIE_STATUS`
        2. Clear the buffer available bit in the buffer status
        3. Toggle the `next_pid` bit in the `hw_endpoint`structure so that the next time the buffer is set up, it has the correct PID.
        4. Notify the host stack that the transfer completed with 0 data transferred. This will clear the busy bit for the endpoint.
        5. Restore the `INTE` register and return;
    8. After the USB host stack task gets the xfer complete notification, the IN endpoint busy status will be clear. This will allow the OUT endpoint to use EPx.

## Code

```
// This function is a hack. It appears that the RP2040 host
// controller will keep retrying a NAK'd bulk transfer forever
// until the transfer completes. This function detects that
// the sie status register NAK bit is set, stops the
// current transaction, clears the NAK status, and fixes
// up the endpoint so we can safely start a new transaction
void hcd_edpt_clear_in_on_nak(uint8_t dev_addr, uint8_t ep_addr)
{
    struct hw_endpoint *ep = get_dev_ep(dev_addr, ep_addr);
    if (ep == _current_epx_endpoint)
    {
        if (nak_received() && (usb_hw->ints & (USB_INTS_BUFF_STATUS_BITS | USB_INTS_TRANS_COMPLETE_BITS)) == 0)
        {
            if (skip_hcd_edpt_clear_in_on_nak > 0)
            {
                --skip_hcd_edpt_clear_in_on_nak;
                clear_nak_received();
                return;
            }
            // Disable USB interrupts in case there is an in-flight transaction
            uint32_t temp_inte = usb_hw->inte;
            usb_hw->inte = 0;
            // clear the NAK status bit
            clear_nak_received();
            // Wait for either the next nak or any raw irq that would interrupt the USB
            // The amount of time between host NAK retries is set in the nak_poll register
            uint32_t masked_ints = (usb_hw->intr & temp_inte);
            while (!nak_received() && masked_ints == 0)
                masked_ints = (usb_hw->intr & temp_inte);
            // TODO: Enter critical section (disable all IRQs) here?
            if (masked_ints == 0 && nak_received())
            {
                // stop the current transaction to free up the host epx HW
                usb_hw_set->sie_ctrl = USB_SIE_CTRL_STOP_TRANS_BITS;
                // TODO: Exit critical section here?
                // clear the NAK status bit
                clear_nak_received();
                // There was no transfer, but the logic that sets up the buffer
                // control register needs to see USB_BUF_CTRL_AVAIL bit clear
                // or else it will assert.
                _hw_endpoint_buffer_control_clear_mask32(ep, USB_BUF_CTRL_AVAIL);
                // Fixup next PID: prepare_ep_buffer() toggles the PID, but there was
                // no successful data transfer, so we need to toggle it here so that
                // it is back where we started.
                ep->next_pid ^= 1u;
                // Notify host stack that the transfer is done (clears busy)
                hw_xfer_complete(ep, XFER_RESULT_SUCCESS);
            }
            // restore the USB interrupts; a transaction is in process
            usb_hw->inte = temp_inte;
            // TODO: Exit critical section here if still in one?
        }
    }
}
```

# Other Issues

1. The algorithm to abort the IN transfer should not be attempted when a device first connects. I am not sure why that is, but it may have to do with the device not fully ready for bulk transfers after reset or power-up.
2. The default interval between auto-retry after NAK is 10 microseconds. I didn't characterize how long it takes between the time NAK is received and the `STOP_TRANS` bit in the `SIE_CTRL` register stops the transfer. To be safe, I set the `NAK_POLL` register to retry every 100 microseconds instead. It seems to work in my testing, but it probably should be characterized more carefully.
3. The MIDI Host code I wrote was not tested with a hub. At the time I worked on this, tinyusb hub support does not function on the RP2040. I do not know if the `SIE_STATUS` `NAK_REC` bit can be set if an interrupt endpoint responds with NAK. If it can, there is no hope to make this scheme work with hubs because hub status uses an interrupt endpoint.
4. I found that if the host code set the `START_TRANS` bit in the `SIE_CTRL` at the same time as all other other bits to start a bulk transfer, sometimes the transfer would not complete. The work-around was to write the other bits in `SIE_CTRL` first and then set `START_TRANS`

**Comment by conkerkh** [comment]
I know it's an old topic but I would suggest that you check if your DATA0/1 PID order is correct, I believe that the current host pico driver layer implementation will not remember DATA pid of the endpoint when switching between endpoints on EPx. This is actually a pretty big issue, NAK will be usually issued when the device is busy doing sth and USB Controller will retry until it can finalize the transaction. Try to connect an MSC device and check if you can get the data about flash driver (capacity). It's a similar case, and the host PICO will stay stucked waiting for data sending NAKed packets.

**Comment by rppicomidi** [comment]
@conkerkh Thanks for the reply. The changes I made to hcd_rp2040.c in my pull request allocate a new hw endpoint structure for each bulk endpoint. The toggle bit is tracked there. It seems right, but I don't have a USB Protocol Analyzer to check for sure. The code I have seems to work OK for me and for other people except if the transfer size is not 64 bytes. There was no bulk endpoint support for the RP2040 when I submitted my pull request, so if MSC is supported now, it is probably implementing bulk transfers differently.

**Comment by rppicomidi** [comment, marked as answer]
There is a solution. See pull request #1434. There seems to be no need to directly share EPx. I must have misread the documentation.

**Comment by wrtlprnft** [comment]
One way that might be less hacky:

When we want to start a bulk IN transfer, set the `EP_CTRL_INTERRUPT_ON_NAK` bit in the epx endpoint control register and raise the values of `NAK_POLL` to the maximum. This means that the retry on NAK has a delay of more than 1 ms, which hopefully means that the interrupt should be able to complete before hardware will retry.

Inside the interrupt handler, if we detect the NAK case, always cancel the retry using `SIE_CTRL.STOP_TRANS`. Then we can decide whether we should try the same transaction, or do another outstanding transaction first, or give up on the transaction.

The real issue is that there seems to be no way for software to query whether the EPX endpoint is currently active. Otherwise, we could just let hardware retry with a shorter `NAK_POLL` value and when we decide the transfer should be cancelled, we could just clear the enable bit of the epx endpoint control register.

**Edit**: Never mind, the “Interrupt on NAK” flag of doesn't exist in host mode. I also don't see another way to get an interrupt when `SIE_STATUS.NAK_REC` is set. That probably means that this idea cannot work :–(
